run-python/python.go

```diff
@@ -0,0 +1,266 @@
+		""
+)
+
+func getPipScript(cache stoic.Cache) (string, error) {
+	var err error
+
+	script, err := ioutil.TempFile("", "get-pip-*.py")
+	if err != nil {
+		return "", fmt.Errorf(
+			"unable to set up temp file for get-pip.py: %v", err)
+	}
+	defer func() {
+		if err != nil {
+			os.Remove(script.Name())
+		}
+		script.Close()
+	}()
+
+	if cacheReader := cache.Get(getPipCacheKey); cacheReader != nil {
+		defer cacheReader.Close()
+
+		_, err = io.Copy(script, cacheReader)
+		if err != nil {
+			return "", fmt.Errorf(
+				"unable to read get-pip.py from cache: %v", err)
+		}
+	} else {
+		resp, err := http.Get(getPipURL)
+		if err != nil {
+			return "", fmt.Errorf(
+				"unable to download get-pip.py script from upstream: %v", err)
+		}
+		defer resp.Body.Close()
+
+		err = cache.Put(getPipCacheKey, io.TeeReader(resp.Body, script))
+		if err != nil {
+			return "", fmt.Errorf(
+				"unable to download get-pip.py script from upstream: %v", err)
+		}
+	}
+
+	return script.Name(), nil
+}
+
+func newPythonEnvironment(root string, python string, cache stoic.Cache) (pythonEnvironment, error) {
+	buf := bytes.NewBuffer(nil)
+	fmt.Fprintf(buf, "# Python: %s\n%s", python, pipRequirements)
+
+	pipCache := filepath.Join(root, "pip-cache")
+	pythonName := filepath.Base(python)
+
+	requirements := buf.Bytes()
+	envHash := sha256.Sum256(requirements)
+	envRoot := filepath.Join(root, fmt.Sprintf("%s-%.4x", pythonName, envHash))
+
+	pe := pythonEnvironment{python, pipCache, envRoot}
+
+	marker := filepath.Join(pe.root, readyBase)
+	if fileExists(marker) {
+		return pe, nil
+	}
+
+	if err := os.MkdirAll(pe.Scripts(), os.ModePerm); err != nil {
+		return pythonEnvironment{}, fmt.Errorf(
+			"unable to setup directory for python environmnent: %v", err)
+	}
+
+	envRequirements := filepath.Join(pe.root, requirementsBase)
+	if err := ioutil.WriteFile(envRequirements, requirements, 0644); err != nil {
+		return pythonEnvironment{}, fmt.Errorf(
+			"unable to write requirements for python environment: %v", err)
+	}
+
+	///...
+	getPip, err := getPipScript(cache)
+	if err != nil {
+		return pythonEnvironment{}, err
+	}
+	defer os.Remove(getPip)
+
+	cmd := exec.Command(python, getPip,
+		"--disable-pip-version-check",
+		"--no-warn-script-location",
+		"--ignore-installed",
+		"--isolated",
+		"--cache-dir", pipCache,
+		pe.installModeForSetup(),
+		"--require-hashes",
+		"--requirement", envRequirements,
+
+		// Disable implicit packages
+		"--no-setuptools", "--no-wheel",
+
+		// HACK: Disable implicit pip
+		// MUST BE LAST parameter. See https://github.com/pypa/pip/issues/3685
+		"--src")
+
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	cmd.Env = pe.environForSetup()
+
+	if err := cmd.Run(); err != nil {
+		return pythonEnvironment{}, fmt.Errorf(
+			"unable to setup pip in python environment: %v", err)
+	}
+
+	ioutil.WriteFile(marker, currentTimestamp(), 0644)
```

Comment:
check err?

Reply:
If we fail to create the marker, the environment is usable for now, but we will attempt to re-create it the next time around... I was thinking we shouldn't fail at this point, but I should double check if re-creating on top is allowed.

I can also add some logging, even if we're ignoring, so I'll fix either way.

I also notice that we may be missing some cleanup in case environment setup fails, a rollback of sorts.

Reply:
Aight, sounds good, in these cases we should explicitly ignore the error with `_ = ...`, once we add a linter it will complain anyway :D